command/metadata.go

```diff
@@ -0,0 +1,125 @@
+package command
+
+import (
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/mackerelio/mackerel-agent/config"
+	"github.com/mackerelio/mackerel-agent/metadata"
+)
+
+func metadataGenerators(conf *config.Config) []*metadata.Generator {
+	generators := make([]*metadata.Generator, 0, len(conf.MetadataPlugins))
+
+	workdir := os.Getenv("MACKEREL_PLUGIN_WORKDIR")
+	if workdir == "" {
+		workdir = os.TempDir()
+	}
+
+	for name, pluginConfig := range conf.MetadataPlugins {
+		generator := &metadata.Generator{
+			Name:     name,
+			Config:   pluginConfig,
+			Tempfile: filepath.Join(workdir, "mackerel-metadata-"+name),
+		}
+		logger.Debugf("Metadata plugin generator created: %#v %#v", generator, generator.Config)
+		generators = append(generators, generator)
+	}
+
+	return generators
+}
+
+type metadataResult struct {
+	namespace string
+	metadata  interface{}
+}
+
+func runMetadataLoop(c *Context, termMetadataCh <-chan struct{}, quit <-chan struct{}) {
+	resultCh := make(chan *metadataResult)
+	for _, g := range c.Agent.MetadataGenerators {
+		go runEachMetadataLoop(g, resultCh, quit)
+	}
+
+	exit := false
+	for !exit {
+		select {
+		case <-time.After(1 * time.Minute):
+		case <-termMetadataCh:
+			logger.Debugf("received 'term' chan for metadata loop")
+			exit = true
+		}
+
+		results := []*metadataResult{}
+		hasResult := true
+		for hasResult {
+			select {
+			case result := <-resultCh:
+				results = append(results, result)
+			default:
+				hasResult = false
+			}
+		}
+
+		for _, result := range results {
+			err := c.API.PutMetadata(c.Host.ID, result.namespace, result.metadata)
+			if err != nil {
+				logger.Errorf("put metadata %q failed: %s", result.namespace, err.Error())
+				// do not retry because huge metadata will be always rejected
```

Comment:
Retry on 5XX errors: https://github.com/mackerelio/mackerel-agent/commit/17a51cbc51becbbcedbbd28e035254349afda83c and avoid infinite retry: https://github.com/mackerelio/mackerel-agent/pull/333/commits/4c6a3d6fbf38ca550d77654f22d0496bb5e56228